tensorflow_probability/python/experimental/mcmc/sample_fold.py

```diff
@@ -0,0 +1,131 @@
+    kernel: An instance of `tfp.mcmc.TransitionKernel` which implements one step
+      of the Markov chain.
+    reducers: A (possibly nested) structure of `Reducer`s to be evaluated
+      on the `kernel`'s samples. If no reducers are given (`reducers=None`),
+      then `None` will be returned in place of streaming calculations.
+    num_burnin_steps: Integer or scalar `Tensor` representing the number
+        of chain steps to take before starting to collect results.
+        Defaults to 0 (i.e., no burn-in).
+    num_steps_between_results: Integer or scalar `Tensor` representing
+      the number of chain steps between collecting a result. Only one out
+      of every `num_steps_between_samples + 1` steps is included in the
+      returned results. Defaults to 0 (i.e., no thinning).
+    parallel_iterations: The number of iterations allowed to run in parallel. It
+      must be a positive integer. See `tf.while_loop` for more details.
+    seed: Optional seed for reproducible sampling.
+    name: Python `str` name prefixed to Ops created by this function.
+      Default value: `None` (i.e., 'mcmc_sample_fold').
+
+  Returns:
+    reduction_results: A (possibly nested) structure of finalized reducer
+      statistics. The structure identically mimics that of `reducers`.
+    end_state: The final state of the Markov chain(s).
+    final_kernel_results: `WithReductionsKernelResults` containing final reducer
+      states. It also includes the kernel results of nested Transition Kernels
+      in its `inner_results` field.
+  """
+  with tf.name_scope(name or 'mcmc_sample_fold'):
+    num_steps = tf.convert_to_tensor(
+        num_steps, dtype=tf.int32, name='num_steps')
+    current_state = tf.nest.map_structure(
+        lambda x: tf.convert_to_tensor(x, name='current_state'),
+        current_state)
+    reducers_was_none = False
+    if reducers is None:
+      reducers = []
+      reducers_was_none = True
+    reduction_kernel = mcmc.WithReductions(
+        inner_kernel=mcmc.SampleDiscardingKernel(
+            inner_kernel=kernel,
+            num_burnin_steps=num_burnin_steps,
+            num_steps_between_results=num_steps_between_results),
+        reducers=reducers,
+    )
+    end_state, final_kernel_results = mcmc.step_kernel(
+        num_steps=num_steps,
+        current_state=current_state,
+        previous_kernel_results=previous_kernel_results,
+        kernel=reduction_kernel,
+        return_final_kernel_results=True,
+        parallel_iterations=parallel_iterations,
+        seed=seed,
+        name=name,
+    )
+    reduction_results = nest.map_structure_up_to(
+        reducers,
+        lambda r, s: r.finalize(s),
+        reducers,
+        final_kernel_results.streaming_calculations,
+        check_types=False)
+    if reduction_results == [] and reducers_was_none:
+      reduction_results = None
+      final_kernel_results = mcmc.WithReductionsKernelResults(
+          None, final_kernel_results.inner_results
```

Comment:
It think we should return the same type as `previous_kernel_results`. Any reason not to?

Reply:
This way, the user can still introspect the kernel results of `WithReductions` and `SampleDiscardingKernel`. The main benefit is that they have a way to re-reference the final reducer states (in `WithReductions`' kernel results). It may be somewhat unintuitive though, but assuming the user has also read the documentation they could also get to the same layer as their given pkr by taking `final_kernel_results.inner_results.inner_results`

Reply:
Hmm, I see the conundrum. Here are my concerns:

1. I kind of dislike the fact that we have `final_kernel_results` and `previous_kernel_results` in the same code block, but with different meanings of `kernel`.
2. Regardless of the name, this will be the first driver which is non-markovian in the sense of thestate type changing; this seems an unfortunate departure from precedent.
3. Allowing people to introspect is nice, but this driver does not let you warmstart the reduction with the reducer state, which is the primary use case I can think of for doing this.

There are a number of things we can try doing to resolve this:

1. Make the driver accept both inner kernel results and outer kernel results. This handles many common use cases, but is ambiguous sometimes.
2. Force the driver to only accept WithReductionKernelResults as input. I think that partially defeats the purpose of this driver, as it was meant to hide that state.
3. Make the driver return the inner kernel results only. This precludes introspection and warmstarting the reducer state, even in principle.
4. Do nothing.

I argue that (3) is the best option, as we can consider this driver to merely be syntax sugar for a common use case, and for people who want to do the introspection you're talking about + warmstarting the reduction state, they should use step_kernel + WithReductions manually (presumably this is the less common use case).

@axch, what do you think?

Reply:
On reflection, I think I agree with (3).  And document that if warm-restart of the reductions is desired, then step_kernel with an explicit WithReductions kernel is the way to go.

(Or, amusingly, sample_fold with no reducers and an explicit WithReductions kernel will also gain the same effect!).

Reasoning: The fundamental concept is step_kernel, and composability through nesting kernels to step.  Sample_fold is a guess as to what people will consider convenient.